File: ru.kmorozov.gbd.core/src/main/java/ru/kmorozov/gbd/core/loader/ListBasedContextLoader.java
package ru.kmorozov.gbd.core.loader;

import ru.kmorozov.db.core.config.IContextLoader;
import ru.kmorozov.db.core.logic.model.book.BookInfo;
import ru.kmorozov.gbd.core.logic.context.IBookListProducer;
import ru.kmorozov.gbd.core.logic.library.LibraryFactory;

import java.util.Set;

public class ListBasedContextLoader implements IContextLoader {

    private final IBookListProducer producer;

    public ListBasedContextLoader(final IBookListProducer producer) {
        this.producer = producer;
    }

    @Override
    public void updateIndex() {

    }

    @Override
    public void updateContext() {

    }

    @Override
    public void updateBookInfo(final BookInfo bookInfo) {

    }

    @Override
    public BookInfo getBookInfo(final String bookId) {
        return LibraryFactory.getMetadata(bookId).getBookExtractor(bookId, null).getBookInfo();
    }

    @Override
    public Set<String> getBookIdsList() {
        return this.producer.getBookIds();
    }

    @Override
    public int getContextSize() {
        return 0;
    }

    @Override
    public void refreshContext() {

    }

    @Override
    public boolean isValid() {
        return true;
    }
}

File: ru.kmorozov.gbd.core/src/main/java/ru/kmorozov/gbd/core/loader/ListBasedContextLoader.kt
package ru.kmorozov.gbd.core.loader

import ru.kmorozov.db.core.config.EmptyContextLoader.Companion.EMPTY_CONTEXT_LOADER
import ru.kmorozov.db.core.config.IContextLoader
import ru.kmorozov.db.core.logic.model.book.BookInfo
import ru.kmorozov.gbd.core.logic.context.IBookListProducer
import ru.kmorozov.gbd.core.logic.library.LibraryFactory

class ListBasedContextLoader(private val producer: IBookListProducer) : IContextLoader {

    private val books: MutableMap<String, BookInfo> = HashMap()

    override val empty: Boolean
        get() = false

    override val bookIdsList: Set<String>
        get() = producer.bookIds

    override val contextSize: Int
        get() = producer.bookIds.size

    override val isValid: Boolean
        get() = true

    override fun updateContext() {

    }

    override fun updateBookInfo(bookInfo: BookInfo) {

    }

    override fun getBookInfo(bookId: String): BookInfo {
        if (!books.containsKey(bookId))
            books.put(bookId, LibraryFactory.getMetadata(bookId).getBookInfoExtractor(bookId, EMPTY_CONTEXT_LOADER).bookInfo)

        return books.get(bookId)!!
    }
}
